Fixes #17 : Fixes when set -f option, no set -f option automatically

samsonasik opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Bugfix        | yes

### Description

Re-create https://github.com/zendframework/zend-mail/pull/227 by @tzmfreedom Fixes #17 .  Original PR description : 

```
Zend\Mail\Transport\SendMail always append envelope-from parameter.
But I want to set any envelope-from.
In this PR, only when not set option -f, set -f option automatically.
This means if I set -ffoo@example.com or -bs -ffoo@example.com, Zend\Mail\Transport\Sendmail doesn't set -f option.
```

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

samsonasik committed:
Fixes #17 : Fixes when set -f option, no set -f option automatically

Signed-off-by: Abdul Malik Ikhsan <samsonasik@gmail.com>

weierophinney commented on src/Transport/Sendmail.php:
```diff
@@ -260,6 +260,9 @@ protected function prepareParameters(Mail\Message $message)
         }
 
         $parameters = (string) $this->parameters;
+        if (preg_match('/^\-f.+$/', $parameters) || strpos($parameters, ' -f') !== false) {
```
```suggestion
        if (strpos($parameters, '-f') !== false) {
```

This accomplishes the same thing, with one less operation.

michalbundyra replied:
@weierophinney it doesn't seem right, imo

it's not the same, previously we accept `-f` at the beginning of when it is after a space.

I can imagine case `my-fancy-param` where `-f` is no at the beginning nor after a space.

weierophinney replied:
Then do it as follows:

```php
if (preg_match('/(^| )\-f.+/', $parameters)) {
```

It doesn't make sense to have two conditions when it's essentially the same condition ("is there a -f switch provided?"). The above regex does the "at the beginning" as well as "following a space" conditions.

michalbundyra replied:
that's better ;-) there is also redundant `\` before `-` in the regexp

samsonasik replied:
@weierophinney updated.

weierophinney reviewed (changes requested):
I've made one suggestion to how you check for the `-f` flag.

Additionally, we need tests demonstrating that the `setParameters()` method uses the parameter passed to the transport and NOT the sender and/or from address when it is provided; currently, you only demonstrate it is used when provided, and not used when not present, but not the additional case where it is not provided and falls through to the sender or from address.

samsonasik committed:
use regex that check  "at the beginning" as well as "following a space" conditions

Signed-off-by: Abdul Malik Ikhsan <samsonasik@gmail.com>

weierophinney committed:
refactor: simplify testing for -f switch via data provider

This patch combines what was previously two test methods,
`testNotSetOptionAutomaticallyOnLeadingF` and
`testNotSetOptionAutomaticallyOnMiddleF`, into one method that better
describes the behaviour (`testDoesNotInjectFromParameterFromSenderWhenFromOptionPresentInParameters`)
and which uses a data provider to demonstrate various permutations of
parameters containing a `-f` switch.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: adds CHANGELOG entry for #104

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney reviewed (approved):
I've combined the two unit tests into a single one backed by a data provider. I think it's ready to :ship: once tests have run.
